### PR TITLE
Dev test comprehension (PROGRESS PR)

### PR DESCRIPTION
**Pull request by RyanDJLee**
- Set environment for ListComp nodes (DictComp, SetComp to be implemented)
- Implement visitor for Comprehension node.
- Implement visitor for ListComp node.
- Add stub test cases for ListComp  node.

### CODE REVIEW

File: python_ta/transforms/type_inference_visitor.py [hunk 1]
```diff
@@ -63,6 +64,14 @@ def _set_function_def_environment(self, node):
         self._populate_local_env(node)
         node.type_environment.locals['return'] = self.type_constraints.fresh_tvar()
 
+    def _set_listcomp_environment(self, node):
+        """Set the environment of a ListComp node representing a list
+        comprehension expression."""
+        node.type_environment = Environment()
+        for name in node.locals:
+            node.type_environment.locals[name] = self.\
```
Comment (david-yz-liu) [inline python_ta/transforms/type_inference_visitor.py]: Start the `self` on a new line.

File: python_ta/transforms/type_inference_visitor.py [hunk 2]
```diff
@@ -290,6 +299,32 @@ def visit_ifexp(self, node):
         else:
             node.type_constraints = TypeInfo(Any)
 
+    def visit_comprehension(self, node):
+        if isinstance(node.iter, Name):
```
Comment (david-yz-liu) [inline python_ta/transforms/type_inference_visitor.py]: This special case shouldn't be necessary. The `Name` node should have as its type a `TypeVar` that you can look up.

File: tests/test_type_inference/test_comprehension.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import nose
+from hypothesis import given, settings, assume
+from typing import Callable, Any
+import tests.custom_hypothesis_support as cs
+settings.load_profile("pyta")
+
+
+# def test_list_comprehension1():
+#     """Test Comprehension node visitor representing a comprehension expression."""
+#     program = f'a = [num for num in [1,2,3]]'
```
Comment (david-yz-liu) [inline tests/test_type_inference/test_comprehension.py]: If you don't have the simple one working then you shouldn't be working on the more complex one.

File: python_ta/transforms/type_inference_visitor.py [hunk 4]
```diff
@@ -290,6 +299,32 @@ def visit_ifexp(self, node):
...
+            target_tvar = node.parent.type_environment.lookup_in_env(node.target.name)
+            self.type_constraints.unify(target_tvar, rtype.__args__[0])
+        node.type_constraints = TypeInfo(NoType)
+
+    def visit_listcomp(self, node):
+        # based on the types of the generators.. can set up the type constraints.
+        # the type of the list comprehension should be a list of the elt's types
+        # if the elt is a name node, look up the name in the listcomp node's env and set the type const as that
+        # if it is an operation node,
+        if isinstance(node.elt, Name):
```
Comment (david-yz-liu) [inline python_ta/transforms/type_inference_visitor.py]: Again, you are treating `Name` as a special case. You should not do this.

File: python_ta/transforms/type_inference_visitor.py [hunk 5]
```diff
@@ -290,6 +298,28 @@ def visit_ifexp(self, node):
...
+            self.type_constraints.unify(target_tvar, rtype.__args__[0])
+        node.type_constraints = TypeInfo(NoType)
+
+    def visit_listcomp(self, node):
+        # based on the types of the generators.. can set up the type constraints.
+        # the type of the list comprehension should be a list of the elt's types
+        # if the elt is a name node, look up the name in the listcomp node's env and set the type const as that
+        # if it is an operation node,
+        node.type_constraints = TypeInfo(List[self.type_constraints.lookup_concrete(node.type_environment.
+                                                                                        lookup_in_env(node.elt.name))])
```
Comment (david-yz-liu) [inline python_ta/transforms/type_inference_visitor.py]: You still aren't thinking recursively here. You should be accessing `node.elt.type_constraints.type`.

Comment (RyanDJLee) [inline python_ta/transforms/type_inference_visitor.py]: Yes, I realized this. I should not have pushed this change

**Pull request closed without merging**